cdap-spark-core-base/src/main/java/io/cdap/cdap/app/runtime/spark/SparkPackageUtils.java

```diff
@@ -462,27 +469,57 @@ private static SparkFramework prepareSparkFramework(SparkCompat sparkCompat,
     String archiveName = "spark.archive-" + sparkVersion + "-" + VersionInfo.getVersion() + ".zip";
     Location frameworkDir = locationFactory.create("/framework/spark");
     Location frameworkLocation = frameworkDir.append(archiveName);
-
-    if (!frameworkLocation.exists()) {
-      File archive = new File(tempDir, archiveName);
-      try {
-        try (ZipOutputStream zipOutput = new ZipOutputStream(new BufferedOutputStream(new FileOutputStream(archive)))) {
-          zipOutput.setLevel(Deflater.NO_COMPRESSION);
-          for (File file : getLocalSparkLibrary(sparkCompat)) {
-            zipOutput.putNextEntry(new ZipEntry(file.getName()));
-            Files.copy(file.toPath(), zipOutput);
-            zipOutput.closeEntry();
+    Location writeInProgressLoc = frameworkDir.append("write_in_progress");
+
+    //Create writing file
+    boolean isWriteComplete = false;
+    while (!isWriteComplete) {
+      if (writeInProgressLoc.createNew()) {
```

Comment:
You don't need to lock the location this early. You can create the zip file locally first (in a local tmp file) and only call `createNew` when you are about to upload the file.